ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/authorization/AsyncForwardingListener.scala

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.server.api.authorization
+
+import io.grpc.ServerCall
+
+/** This listener buffers all messages until `setNextListener` is called,
+  * at which point all buffered messages are sent to the given listener.
+  * From then on, all future messages are sent directly to the given listener.
+  *
+  * The target listener is usually created through `Contexts.interceptCall` or `ServerCallHandler.startCall`.
+  * */
+abstract class AsyncForwardingListener[ReqT] extends ServerCall.Listener[ReqT] {
+  protected type Listener = ServerCall.Listener[ReqT]
+  private var stash: List[Listener => Unit] = List.empty
+  private var nextListener: Option[Listener] = None
+
+  private def enqueueOrProcess(msg: Listener => Unit): Unit = this.synchronized {
+    nextListener.fold[Unit]({stash = stash :+ msg})(l => msg(l))
```

Comment:
Given the heavily procedural nature of this piece of code, it probably makes sense in this case to use a more procedural idiom:
```scala
      if (nextListener.isDefined) {
        msg(nextListener.get)
      } else {
        stash :+ msg
      }
```